expr/http_endpoint.go

```diff
@@ -660,44 +660,67 @@ func (e *HTTPEndpointExpr) validateParams() *eval.ValidationErrors {
 // validateHeaders makes sure headers are of an allowed type and the method
 // payload contains the headers.
 func (e *HTTPEndpointExpr) validateHeaders() *eval.ValidationErrors {
-	headers := AsObject(e.Headers.Type)
-	if len(*headers) == 0 {
-		return nil
-	}
 	verr := new(eval.ValidationErrors)
-	for _, nat := range *headers {
-		if IsObject(nat.Attribute.Type) {
-			verr.Add(e, "header %s cannot be an object, header type must be primitive or array", nat.Name)
-		} else if arr := AsArray(nat.Attribute.Type); arr != nil {
-			if !IsPrimitive(arr.ElemType.Type) {
-				verr.Add(e, "elements of array header %s must be primitive", nat.Name)
+
+	// We have to figure out the actual type for the params because the actual
+	// type is initialized only during the finalize phase. In the validation
+	// phase, all param types are string type by default unless specified
+	// expliclty.
+	init := func(ma *MappedAttributeExpr) *MappedAttributeExpr {
+		ma = DupMappedAtt(ma)
```

Comment:
Can this be encapsulated in a private function that is used by both the validation and finalization steps?